kyo-core/shared/src/main/scala/kyo/Result.scala

```diff
@@ -0,0 +1,140 @@
+package kyo
+
+import Result.*
+import kyo.Flat
+import scala.util.control.NonFatal
+
+opaque type Result[+T] >: (Success[T] | Failure[T]) = Success[T] | Failure[T]
+
+object Result:
+    import internal.*
+
+    given [T: Flat]: Flat[Result[T]]                                = Flat.unsafe.bypass
+    given [T](using CanEqual[T, T]): CanEqual[Result[T], Result[T]] = CanEqual.derived
+
+    def apply[T](expr: => T): Result[T] =
+        try
+            Success(expr)
+        catch
+            case ex: Throwable if NonFatal(ex) => Failure(ex)
+
+    def success[T](value: T): Result[T] = Success(value)
+
+    def failure[T](exception: Throwable): Result[T] = Failure(exception)
+
+    def fromEither[T](either: Either[Throwable, T]): Result[T] = either match
+        case Right(value)    => Success(value)
+        case Left(exception) => Failure(exception)
+
+    opaque type Success[+T] = T | SuccessFailure[T]
+
+    object Success:
+
+        def apply[T](value: T): Success[T] =
+            value match
+                case v: SuccessFailure[?]     => v.nest.asInstanceOf[Success[T]]
+                case v: Failure[T] @unchecked => SuccessFailure(v)
+                case v                        => v
+
+        // TODO avoid Option allocation
+        def unapply[T](self: Result[T]): Option[T] =
+            (self: @unchecked) match
+                case _: Failure[?] => None
+                case s: SuccessFailure[?] =>
+                    Some(s.unnest.asInstanceOf[T])
+                case s: T @unchecked => Some(s)
+
+    end Success
+
+    case class Failure[+T](exception: Throwable)
```

Comment:
Assuming this will be the backing for `Fiber`, do we want to store other information? Trace? Do we need a `Cause` type like ZIO?

Reply:
Yes, I'm planning to use it to avoid `T < IOs` in fiber completions. The `Option`/`Maybe` was a little detour to explore how to support nesting without allocations :)

I think we could keep the trace information in the exception itself since we can decide which exceptions benefit from a origin `Trace` and which don't.